sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctAttributesVisitor.scala

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeSet, ExpressionSet, NamedExpression}
+import org.apache.spark.sql.catalyst.plans.{LeftAnti, LeftSemi}
+
+/**
+ * A visitor pattern for traversing a [[LogicalPlan]] tree and propagate the distinct attributes.
+ */
+object DistinctAttributesVisitor extends LogicalPlanVisitor[Seq[AttributeSet]] {
+
+  private def allDistinctAttributes(
+      expressions: Seq[NamedExpression],
+      distinctAttributes: Seq[ExpressionSet]): Seq[AttributeSet] = {
+    expressions.toSet.subsets(distinctAttributes.map(_.size).min).filter { s =>
```

Comment:
This looks quite inefficient, to generate all the subsets. I think the logic here is:
1. produce correct distinct keys by mapping expressions to attributes w.r.t. project list
2. expand the distinct keys according to the aliases
3. prune invalid distinct keys that are not output by the project list.

Reply:
Add a filter before build subsets:
```scala
    val expressions = keys.flatMap(_.toSet)
    projectList.filter {
      case a: Alias => expressions.exists(_.semanticEquals(a.child))
      case ne: NamedExpression => expressions.exists(_.semanticEquals(ne))
    }.toSet.subsets(keys.map(_.size).min).filter { s =>
      val references = s.map {
        case a: Alias => a.child
        case ne => ne
      }
      keys.exists(_.equals(ExpressionSet(references)))
    }.map(s => AttributeSet(s.map(_.toAttribute))).toSet
```